#  This is second step of World t20 Cup Score Prediction  

In [10]:
import pandas as pd
import pickle
import numpy as np
import sklearn

In [4]:
# loading the dataframe which is formed from last extraction work

In [7]:
!pip3 install scikit-learn==0.24.dev0

In [11]:
sklearn. __version__

'0.24.1'

In [2]:
df = pickle.load(open('dataset_level2.pkl','rb'))

In [3]:
df



match_id batting_team bowling_team  ball  runs player_dismissed     city  \
0           2    Australia    Sri Lanka   0.1     0                0      NaN   
1           2    Australia    Sri Lanka   0.2     0                0      NaN   
2           2    Australia    Sri Lanka   0.3     1                0      NaN   
3           2    Australia    Sri Lanka   0.4     2                0      NaN   
4           2    Australia    Sri Lanka   0.5     0                0      NaN   
..        ...          ...          ...   ...   ...              ...      ...   
121       964    Sri Lanka    Australia  19.3     1                0  Colombo   
122       964    Sri Lanka    Australia  19.4     0                0  Colombo   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva  Colombo   
124       964    Sri Lanka    Australia  19.6     2                0  Colombo   
125       964    Sri Lanka    Australia  19.7     1                0  Colombo   

                        venue  
0    Melbourne Cricket Ground  
1    Melbourne Cricket Ground  
2    Melbourne Cricket Ground  
3    Melbourne Cricket Ground  
4    Melbourne Cricket Ground  
..                        ...  
121       R Premadasa Stadium  
122       R Premadasa Stadium  
123       R Premadasa Stadium  
124       R Premadasa Stadium  
125       R Premadasa Stadium  

[63888 rows x 8 columns]

### List of columns that are of our interest  

        # batting_team
        # bowling team
        # city
        # current score
        # ball left
        # wickets left
        # current run rate
        # last five

In [5]:
# Checking null values in the existing dataframe

In [6]:
df.isnull().sum()    # here city column is havung 8548 null values

match_id               0
batting_team           0
bowling_team           0
ball                   0
runs                   0
player_dismissed       0
city                8548
venue                  0
dtype: int64

In [7]:
# city column is havung maximum number of null values


In [11]:
df[df['city'].isnull()]['venue'].value_counts()     #GIVES INFORMATION ABOUT VENUE WHERE CITY VALUE IS NULL

#df['venue'].value_counts()    # TELL US ABOUT VENUE INFORMATION IN THE DATAFRAME

Dubai International Cricket Stadium        2969
Pallekele International Cricket Stadium    2066
Melbourne Cricket Ground                   1453
Sydney Cricket Ground                       749
Adelaide Oval                               498
Harare Sports Club                          372
Sharjah Cricket Stadium                     249
Sylhet International Cricket Stadium        128
Carrara Oval                                 64
Name: venue, dtype: int64

# SPLITTING VENUE NAME WHERE CITY VALUE IS NULL

In [12]:
cities = np.where(df['city'].isnull(),df['venue'].str.split().apply(lambda x:x[0]),df['city'])


# NP.WHERE HAS 3 PARAMETER (condition, 1t option if it true, 2nd option if it true)

#### Here we are checking where city value is null 

#### if value is null then venue will have operation of spliting and 1st position will be taken into account

#### else non null value will be used


#### values stored in cities as array

In [13]:
cities

array(['Melbourne', 'Melbourne', 'Melbourne', ..., 'Colombo', 'Colombo',
       'Colombo'], dtype=object)

In [14]:
df['city'] = cities

In [15]:
df.isnull().sum()     # no null values 

match_id            0
batting_team        0
bowling_team        0
ball                0
runs                0
player_dismissed    0
city                0
venue               0
dtype: int64

In [16]:
df.drop(columns=['venue'],inplace=True)     # deleting venue column as it is of no values in future

In [17]:
df  #city can be seen splitted from venue column

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  
0    Melbourne  
1    Melbourne  
2    Melbourne  
3    Melbourne  
4    Melbourne  
..         ...  
121    Colombo  
122    Colombo  
123    Colombo  
124    Colombo  
125    Colombo  

[63888 rows x 7 columns]

# Here we have 63888 rows 



# we will only consider city which has only hosted more than 5 matches



   # In Order to do above, we will check condition ie greater than 600ball   

In [18]:
eligible_cities = df['city'].value_counts()[df['city'].value_counts() > 600].index.tolist()

In [19]:
df = df[df['city'].isin(eligible_cities)]       # Those who satisfy condition of matching will be only stored

In [20]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  
0    Melbourne  
1    Melbourne  
2    Melbourne  
3    Melbourne  
4    Melbourne  
..         ...  
121    Colombo  
122    Colombo  
123    Colombo  
124    Colombo  
125    Colombo  

[50501 rows x 7 columns]

# Current Score is calcualted by cummaltive summation

   # 1. groupby columns
   # 2. use cumsum() function to do cummaltive sum of runs

In [21]:
df['current_score'] = df.groupby('match_id').cumsum()['runs']

<ipython-input-21-ac174c139314>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['current_score'] = df.groupby('match_id').cumsum()['runs']


In [22]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  current_score  
0    Melbourne              0  
1    Melbourne              0  
2    Melbourne              1  
3    Melbourne              3  
4    Melbourne              3  
..         ...            ...  
121    Colombo            125  
122    Colombo            125  
123    Colombo            125  
124    Colombo            127  
125    Colombo            128  

[50501 rows x 8 columns]

# Calculate "ball_no"



# Steps 

   # 1. Divide over column by spliting based on dot(.) Value
   
   # Ex- 4.3over is divided into '4' over and '3' ball

In [25]:
df['over'] = df['ball'].apply(lambda x:str(x).split(".")[0])
df['ball_no'] = df['ball'].apply(lambda x:str(x).split(".")[1])
#df

<ipython-input-25-4b49976b94f5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['over'] = df['ball'].apply(lambda x:str(x).split(".")[0])
<ipython-input-25-4b49976b94f5>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['ball_no'] = df['ball'].apply(lambda x:str(x).split(".")[1])


In [26]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  current_score over ball_no  
0    Melbourne              0    0       1  
1    Melbourne              0    0       2  
2    Melbourne              1    0       3  
3    Melbourne              3    0       4  
4    Melbourne              3    0       5  
..         ...            ...  ...     ...  
121    Colombo            125   19       3  
122    Colombo            125   19       4  
123    Colombo            125   19       5  
124    Colombo            127   19       6  
125    Colombo            128   19       7  

[50501 rows x 10 columns]

# Calculated balls_bowled



   # Multiply 'over' column by 6 and add 'ball no' (Convert into Integer)
   
   
   
   
   # Example - 0.5 over is convert into 5balls bowled

In [27]:
df['balls_bowled'] = (df['over'].astype('int')*6) + df['ball_no'].astype('int')
df

<ipython-input-27-e7d17f656852>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['balls_bowled'] = (df['over'].astype('int')*6) + df['ball_no'].astype('int')


match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  current_score over ball_no  balls_bowled  
0    Melbourne              0    0       1             1  
1    Melbourne              0    0       2             2  
2    Melbourne              1    0       3             3  
3    Melbourne              3    0       4             4  
4    Melbourne              3    0       5             5  
..         ...            ...  ...     ...           ...  
121    Colombo            125   19       3           117  
122    Colombo            125   19       4           118  
123    Colombo            125   19       5           119  
124    Colombo            127   19       6           120  
125    Colombo            128   19       7           121  

[50501 rows x 11 columns]

# Calculating balls left 

# Simply sub 120 from balls bowled 
# and 
# Do simple Operation for negative values

In [30]:
df['balls_left'] = 120 - df['balls_bowled']
df['balls_left'] = df['balls_left'].apply(lambda x:0 if x<0 else x)
#df

<ipython-input-30-f4e4f23c8bc5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['balls_left'] = 120 - df['balls_bowled']
<ipython-input-30-f4e4f23c8bc5>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['balls_left'] = df['balls_left'].apply(lambda x:0 if x<0 else x)


In [29]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  current_score over ball_no  balls_bowled  balls_left  
0    Melbourne              0    0       1             1         119  
1    Melbourne              0    0       2             2         118  
2    Melbourne              1    0       3             3         117  
3    Melbourne              3    0       4             4         116  
4    Melbourne              3    0       5             5         115  
..         ...            ...  ...     ...           ...         ...  
121    Colombo            125   19       3           117           3  
122    Colombo            125   19       4           118           2  
123    Colombo            125   19       5           119           1  
124    Colombo            127   19       6           120           0  
125    Colombo            128   19       7           121           0  

[50501 rows x 12 columns]

# CALCULATE 'player dismissed' and 'wicket left'



# Steps

#   Convert player's name into '1' if present , else '0' 

   # Convert into integer format

   # Do Group by based on match id
   
   # Apply cumsum()[ie cummaltive summation] 
   
# And for 'wicket left ' do Subtract  'player dismissed' from 10

In [31]:
df['player_dismissed'] = df['player_dismissed'].apply(lambda x:0 if x=='0' else 1)
df['player_dismissed'] = df['player_dismissed'].astype('int')
df['player_dismissed'] = df.groupby('match_id').cumsum()['player_dismissed']
df['wickets_left'] = 10 - df['player_dismissed']

<ipython-input-31-e0234112d42f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['player_dismissed'] = df['player_dismissed'].apply(lambda x:0 if x=='0' else 1)
<ipython-input-31-e0234112d42f>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['player_dismissed'] = df['player_dismissed'].astype('int')
<ipython-input-31-e0234112d42f>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in th

In [32]:
df

match_id batting_team bowling_team  ball  runs  player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                 0   
1           2    Australia    Sri Lanka   0.2     0                 0   
2           2    Australia    Sri Lanka   0.3     1                 0   
3           2    Australia    Sri Lanka   0.4     2                 0   
4           2    Australia    Sri Lanka   0.5     0                 0   
..        ...          ...          ...   ...   ...               ...   
121       964    Sri Lanka    Australia  19.3     1                 8   
122       964    Sri Lanka    Australia  19.4     0                 8   
123       964    Sri Lanka    Australia  19.5     0                 9   
124       964    Sri Lanka    Australia  19.6     2                 9   
125       964    Sri Lanka    Australia  19.7     1                 9   

          city  current_score over ball_no  balls_bowled  balls_left  \
0    Melbourne              0    0       1             1         119   
1    Melbourne              0    0       2             2         118   
2    Melbourne              1    0       3             3         117   
3    Melbourne              3    0       4             4         116   
4    Melbourne              3    0       5             5         115   
..         ...            ...  ...     ...           ...         ...   
121    Colombo            125   19       3           117           3   
122    Colombo            125   19       4           118           2   
123    Colombo            125   19       5           119           1   
124    Colombo            127   19       6           120           0   
125    Colombo            128   19       7           121           0   

     wickets_left  
0              10  
1              10  
2              10  
3              10  
4              10  
..            ...  
121             2  
122             2  
123             1  
124             1  
125             1  

[50501 rows x 13 columns]

# Calculate 'CRR' 


                #using formula  crr=current_score*6/balls bowled  

In [33]:
df['crr'] = (df['current_score']*6)/df['balls_bowled']

<ipython-input-33-fad47fc85776>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['crr'] = (df['current_score']*6)/df['balls_bowled']


In [34]:
df

match_id batting_team bowling_team  ball  runs  player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                 0   
1           2    Australia    Sri Lanka   0.2     0                 0   
2           2    Australia    Sri Lanka   0.3     1                 0   
3           2    Australia    Sri Lanka   0.4     2                 0   
4           2    Australia    Sri Lanka   0.5     0                 0   
..        ...          ...          ...   ...   ...               ...   
121       964    Sri Lanka    Australia  19.3     1                 8   
122       964    Sri Lanka    Australia  19.4     0                 8   
123       964    Sri Lanka    Australia  19.5     0                 9   
124       964    Sri Lanka    Australia  19.6     2                 9   
125       964    Sri Lanka    Australia  19.7     1                 9   

          city  current_score over ball_no  balls_bowled  balls_left  \
0    Melbourne              0    0       1             1         119   
1    Melbourne              0    0       2             2         118   
2    Melbourne              1    0       3             3         117   
3    Melbourne              3    0       4             4         116   
4    Melbourne              3    0       5             5         115   
..         ...            ...  ...     ...           ...         ...   
121    Colombo            125   19       3           117           3   
122    Colombo            125   19       4           118           2   
123    Colombo            125   19       5           119           1   
124    Colombo            127   19       6           120           0   
125    Colombo            128   19       7           121           0   

     wickets_left       crr  
0              10  0.000000  
1              10  0.000000  
2              10  2.000000  
3              10  4.500000  
4              10  3.600000  
..            ...       ...  
121             2  6.410256  
122             2  6.355932  
123             1  6.302521  
124             1  6.350000  
125             1  6.347107  

[50501 rows x 14 columns]

#  Finding runs scored in last 5 overs

In [35]:
groups = df.groupby('match_id')

match_ids = df['match_id'].unique()
last_five = []
for id in match_ids:
    last_five.extend(groups.get_group(id).rolling(window=30).sum()['runs'].values.tolist()) # rolling is used select range of rows to use at once.

In [41]:
#last_five it has list where every match initial 30 balls is nan because we cant have 30 balls played at the starting of innings

In [36]:
df['last_five'] = last_five

<ipython-input-36-83d24d575aa4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['last_five'] = last_five


In [37]:
#df[['batting_team','bowling_team','city','current_score','balls_left','wickets_left','crr','last_five','']]

# Calculating Final Score of the match

In [43]:
final_df = df.groupby('match_id').sum()['runs'].reset_index().merge(df,on='match_id')




#  calculate runs scored based on match id so that match wise score can be calculated

#  reset index() is used making into dataframe

# 2 new rows will be added


Above operation will gives 2 new columns ie---> runs_x and runs_y  


runs_x ----> Total Runs Scored in that Match

runs_y ----> Runs Scored in every ball

In [44]:
final_df

match_id  runs_x batting_team bowling_team  ball  runs_y  \
0             2     168    Australia    Sri Lanka   0.1       0   
1             2     168    Australia    Sri Lanka   0.2       0   
2             2     168    Australia    Sri Lanka   0.3       1   
3             2     168    Australia    Sri Lanka   0.4       2   
4             2     168    Australia    Sri Lanka   0.5       0   
...         ...     ...          ...          ...   ...     ...   
50496       964     128    Sri Lanka    Australia  19.3       1   
50497       964     128    Sri Lanka    Australia  19.4       0   
50498       964     128    Sri Lanka    Australia  19.5       0   
50499       964     128    Sri Lanka    Australia  19.6       2   
50500       964     128    Sri Lanka    Australia  19.7       1   

       player_dismissed       city  current_score over ball_no  balls_bowled  \
0                     0  Melbourne              0    0       1             1   
1                     0  Melbourne              0    0       2             2   
2                     0  Melbourne              1    0       3             3   
3                     0  Melbourne              3    0       4             4   
4                     0  Melbourne              3    0       5             5   
...                 ...        ...            ...  ...     ...           ...   
50496                 8    Colombo            125   19       3           117   
50497                 8    Colombo            125   19       4           118   
50498                 9    Colombo            125   19       5           119   
50499                 9    Colombo            127   19       6           120   
50500                 9    Colombo            128   19       7           121   

       balls_left  wickets_left       crr  last_five  
0             119            10  0.000000        NaN  
1             118            10  0.000000        NaN  
2             117            10  2.000000        NaN  
3             116            10  4.500000        NaN  
4             115            10  3.600000        NaN  
...           ...           ...       ...        ...  
50496           3             2  6.410256       32.0  
50497           2             2  6.355932       32.0  
50498           1             1  6.302521       32.0  
50499           0             1  6.350000       33.0  
50500           0             1  6.347107       32.0  

[50501 rows x 16 columns]

In [45]:
final_df=final_df[['batting_team','bowling_team','city','current_score','balls_left','wickets_left','crr','last_five','runs_x']]

In [46]:
final_df #only 9 columns are of use, Rest are ignored 

batting_team bowling_team       city  current_score  balls_left  \
0        Australia    Sri Lanka  Melbourne              0         119   
1        Australia    Sri Lanka  Melbourne              0         118   
2        Australia    Sri Lanka  Melbourne              1         117   
3        Australia    Sri Lanka  Melbourne              3         116   
4        Australia    Sri Lanka  Melbourne              3         115   
...            ...          ...        ...            ...         ...   
50496    Sri Lanka    Australia    Colombo            125           3   
50497    Sri Lanka    Australia    Colombo            125           2   
50498    Sri Lanka    Australia    Colombo            125           1   
50499    Sri Lanka    Australia    Colombo            127           0   
50500    Sri Lanka    Australia    Colombo            128           0   

       wickets_left       crr  last_five  runs_x  
0                10  0.000000        NaN     168  
1                10  0.000000        NaN     168  
2                10  2.000000        NaN     168  
3                10  4.500000        NaN     168  
4                10  3.600000        NaN     168  
...             ...       ...        ...     ...  
50496             2  6.410256       32.0     128  
50497             2  6.355932       32.0     128  
50498             1  6.302521       32.0     128  
50499             1  6.350000       33.0     128  
50500             1  6.347107       32.0     128  

[50501 rows x 9 columns]

In [ ]:
final_df.isnull().sum() # Here null values will be counted and shown 


#last 5 over runs scored will have Maximum Number of null values 

# will have to droped them all

In [47]:
final_df.dropna(inplace=True)

<ipython-input-47-3665e38311c9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df.dropna(inplace=True)


In [48]:
final_df.isnull().sum() # no columns have any nulll values

batting_team     0
bowling_team     0
city             0
current_score    0
balls_left       0
wickets_left     0
crr              0
last_five        0
runs_x           0
dtype: int64

# Shuffling ---> In Order to prevent from Biased , we will be doing uniform Random sampling in the Dataframe

In [49]:
final_df = final_df.sample(final_df.shape[0])

In [50]:
final_df.sample(2)

batting_team bowling_team      city  current_score  balls_left  \
31894        India      England   Colombo             54          82   
19680      England  West Indies  Trinidad             42          86   

       wickets_left       crr  last_five  runs_x  
31894             9  8.526316       40.0     170  
19680             9  7.411765       37.0     121

# From Here

# Only MACHINE LEARNING TASK

In [ ]:
# Dividing data into train and test 



#TRAIN = 80%
#TEST = 20%

In [51]:
X = final_df.drop(columns=['runs_x'])
y = final_df['runs_x']
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=1)

In [52]:
X_train

batting_team bowling_team        city  current_score  balls_left  \
7025   South Africa    Sri Lanka   Centurion             94          54   
23697     Australia     Pakistan   Melbourne            103          28   
24994   New Zealand     Pakistan    Barbados             43          81   
19366  South Africa    Australia   Centurion             95          34   
21797     Sri Lanka  West Indies      London             50          72   
...             ...          ...         ...            ...         ...   
22970     Australia     Pakistan       Dubai            100          17   
37537   West Indies      England    Barbados            136          16   
12405      Pakistan      England  Manchester            156          21   
50473     Sri Lanka    Australia     Colombo            100          23   
35094         India      England      Mumbai             99          42   

       wickets_left       crr  last_five  
7025              9  8.545455       55.0  
23697             3  6.717391       27.0  
24994             8  6.615385       24.0  
19366             7  6.627907       27.0  
21797            10  6.250000       38.0  
...             ...       ...        ...  
22970             2  5.825243       16.0  
37537             7  7.846154       43.0  
12405             7  9.454545       52.0  
50473             3  6.185567       27.0  
35094             6  7.615385       28.0  

[30781 rows x 8 columns]

In [13]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.metrics import r2_score,mean_absolute_error

In [57]:
#installing xgboost if module is not found


# import sys
# !{sys.executable} -m pip install xgboost

# Transformation of character into integer using one hot encoding

# Rest Integer are left passed

In [59]:
trf = ColumnTransformer([
    ('trf',OneHotEncoder(sparse=False,drop='first'),['batting_team','bowling_team','city'])
]
,remainder='passthrough')

# Defining Pipeline of Sklearn



# 1st Step ----> Transformation
# 2nd Step-----> Standard Scaling
# 3rd Step -----> Using XG Boost Regressor With Changing Hyper-Parameter


In [60]:
pipe = Pipeline(steps=[
    ('step1',trf),
    ('step2',StandardScaler()),
    ('step3',XGBRegressor(n_estimators=1000,learning_rate=0.2,max_depth=12,random_state=1))
])

In [64]:
pipe.fit(X_train,y_train)                           # Training Model 
y_pred = pipe.predict(X_test)                       # Prediction of Model

print(r2_score(y_test,y_pred))
print(mean_absolute_error(y_test,y_pred))

0.98724139046809
1.6419742642717896


In [65]:
pickle.dump(pipe,open('pipe.pkl','wb'))

In [63]:
# import xgboost
# xgboost.__version__

'1.7.1'